File: 0567-permutation-in-string/0567-permutation-in-string.py
from collections import defaultdict

class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        if len(s1) > len(s2):
            return False
        
        s1CharCount = defaultdict(int)
        for i in range(len(s1)):
            s1CharCount[s1[i]] += 1

        subStringCharCount = defaultdict(int)
        for i in range(len(s1)):
            subStringCharCount[s2[i]] += 1

        if subStringCharCount == s1CharCount:
            return True

        l = 0
        for r in range(len(s1), len(s2)):
            subStringCharCount[s2[r]] += 1
            subStringCharCount[s2[l]] -= 1
            filtered_dict = {key: value for key, value in subStringCharCount.items() if value >= 1}
            l += 1
            if filtered_dict == s1CharCount:
                return True

        return False
                

File: 0567-permutation-in-string/0567-permutation-in-string.java
class Solution {
    public boolean checkInclusion(String s1, String s2) {
        if(s1.length() > s2.length()) return false;
        
        
        int [] s1Count = new int[26];
        int [] s2Count = new int[26];
        
        for(int i=0; i<s1.length(); i++){
            s1Count[s1.charAt(i)-'a'] += 1;
            s2Count[s2.charAt(i)-'a'] += 1;
        }
        
        for(int i=s1.length(); i<s2.length(); i++){
            if(this.containsPermutaition(s1Count, s2Count)) return true;
            
            s2Count[s2.charAt(i)-'a'] += 1;
            s2Count[s2.charAt(i-s1.length())-'a'] -= 1;
        }
        
        if(this.containsPermutaition(s1Count, s2Count)) return true;
        return false;
        
    }
    
    
    private boolean containsPermutaition(int[] c1, int[] c2){
        for(int i=0; i<26; i++){
            if(c1[i]!=0 && c1[i]!=c2[i]) return false;
        }
        return true;
    }
}